### PR TITLE
Fixed bug in duration parsing (https://github.com/eilslabs/BatchEupho…

### PR DESCRIPTION
**Pull request by vinjana**
…ria/issues/55). Add BEException class.

### CODE REVIEW

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 1]
```diff
@@ -730,70 +740,74 @@ class PBSJobManager extends ClusterJobManager<PBSCommand> {
...
+                if (jobResult.get("start_time")) // The timepoint the job was started.
+                    gj.setStartTime(parseTime(jobResult.get("start_time")))
+                if (jobResult.get("comp_time"))  // The timepoint the job was completed.
+                    gj.setEndTime(parseTime(jobResult.get("comp_time")))
+                if (jobResult.get("etime"))  // The time that the job became eligible to run, i.e. in a queued state while residing in an execution queue.
+                    gj.setEligibleTime(parseTime(jobResult.get("etime")))
+
+                return queriedExtendedStates.put(it.getKey(), gj)
+
+            } catch (Exception ex) {
```
Comment (dankwart-de) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: Just querying a general exception and throwing a BEException looks a bit like: "I don't know what I am doing so I just do something". Do you know some of the possible exceptions? If so catch them and do proper messages.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 2]
```diff
@@ -696,31 +688,49 @@ class PBSJobManager extends ClusterJobManager<PBSCommand> {
...
-        } as Map<String, Map<String, String>>
+            }.collectEntries { jobId, value ->
+                // join multi-line values
+                value = ((String) value).replaceAll("\n\t", "")
+                [(jobId): value]
+            }.collectEntries { jobId, value ->
+                Map<String, String> p = ((String) value).readLines().collectEntries { String line ->
+                    if (line.startsWith("    ") && line.contains(" = ")) {
+                        String[] parts = line.split(" = ")
+                        [parts.head().replaceAll(/^ {4}/, ""), parts.tail().join(' ')]
```
[the remark this answers is not in the document]
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: Dito. Alignment problem. Please ask Florian.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 3]
```diff
@@ -696,31 +688,49 @@ class PBSJobManager extends ClusterJobManager<PBSCommand> {
...
+                    } else {
+                        [:]
+                    }
+                }
+                [(jobId): p]
+            } as Map<String, Map<String, String>>
+        } catch (Exception ex) {
+            throw new BEException("Error while parsing qstat output:\n${qstatOutput}", ex)
```
[the remark this answers is not in the document]
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: The reason I took a BEException, is that the ParseException does not allow to refer to a causing Exception (the second parameter in this constructor) and I did not want to drop the causing exception. I instead rewrote the code now a bit to make it more robust and dropped the try-catch.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 4]
```diff
@@ -696,31 +688,49 @@ class PBSJobManager extends ClusterJobManager<PBSCommand> {
...
-            Map<String, String> p = ((String) value).readLines().collectEntries { String line ->
-                if (line.startsWith("    ") && line.contains(" = ")) {
-                    String[] parts = line.split(" = ")
-                    [parts.head().replaceAll(/^ {4}/, ""), parts.tail().join(' ')]
+
+        try {
+            return qstatOutput.split(String.format(WITH_DELIMITER, "\n\nJob Id: ")).collectEntries {
+                Matcher matcher = it =~ /^\s*Job Id: (?<jobId>\d+)\..*\n/
+                if (matcher) {
+                    [
```
[the remark this answers is not in the document]
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: I could do this:
```groovy
def result = new HashMap()
if (matcher) {
   result.put(matcher.group("jobId"), it)
} 
result
```

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy (hunk 1 above)
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: Well, it is an improvement over the situation before, without any exception handling. Now at least one learns, what input caused the problem. This is also the reason that I catch and rethrow, just to add context information.

The code is fragile and probably can throw NullPointerExceptions at one point or the other. Do you have any proposals, how I should deal with these?

**Pull request merged**